### PR TITLE
Added health percentage above the heads of players near their names

### PR DESCRIPTION
**Pull request by echo-devim**
The percentage changes its colour based on its level

### CODE REVIEW

File: src/game/game.cpp [hunk 1]
```diff
@@ -25,7 +25,7 @@ namespace game
         player1->version.gpuglslver = glslversion;
         player1->version.crc = versioncrc;
         if(player1->version.branch) delete[] player1->version.branch;
-        player1->version.branch = newstring(versionbranch, MAXBRANCHLEN);
+        player1->version.branch = newstring(versionbranch);
```
Comment (TheAssassin) [inline src/game/game.cpp]: This line of code was correct before, I am not sure why you changed it for this PR. Especially since it does not have anything to do with the PR's purpose. Please undo this change.

**Comment by Calinou** [comment]
I have a few concerns with this pull request:

Most `abovehead*` variables allow you to toggle between displaying information only above teammates, only above enemies, or above both. This one doesn't.
Coloring uses hard-coded discrete steps, it should use continuous coloring (using a simple mathematical formula).

It should also be discussed for balance reasons whether this should be allowed to be seen on enemies or not.

**Comment by echo-devim** [comment]
Another thing, I forgot to use the right buffer size. If you look at the code I used a "large" buffer (health[100]) when it isn't needed.

File: src/game/game.cpp [hunk 2]
```diff
@@ -3168,8 +3169,21 @@ namespace game
         float blend = aboveheadblend*trans;
         if(aboveheadnames && d != player1)
         {
+	    const char *name = colourname(d);
+	    char name_health[strlen(name)+100];
```
Comment (TheAssassin) [inline src/game/game.cpp]: The use of `100` as a constant looks rather random. It is rather easy to calculate how many additional characters are needed.

**Comment by IceflowRE** [comment]
Is it only me or is the intendention wrong now?

File: src/game/game.cpp [hunk 3]
```diff
@@ -3169,18 +3169,22 @@ namespace game
...
-		sprintf(health, "\fs\f[%d]%d%%\fS", health_colour, d->health);
-	        strcat(name_health, health);
+	        const char *name = colourname(d);
+	        int health_length;
+	        int h = d->health;
+	        for (health_length=0; h!=0; health_length++, h=h/10){}
+	        health_length += 17; /* Fixed amount of extra-characters manually calculated */
+	        char name_health[strlen(name)+health_length];
+	        strcpy(name_health, name);
+	        if (aboveheadhealth && (d->health > 0)) {
```
Comment (TheAssassin) [inline src/game/game.cpp]: Please put the brace on a new line.

File: src/game/game.cpp [hunk 4]
```diff
@@ -3169,18 +3169,22 @@ namespace game
...
+		        char health[health_length];
+		        int health_colour = 0x45db1c; /* green */
+		        if (d->health < 30)
+		            health_colour = 0xe03a1d; /* red */
+	            else if (d->health < 70)
```
Comment (TheAssassin) [inline src/game/game.cpp]: Wrong indentation.

File: src/game/game.cpp [hunk 5]
```diff
@@ -3169,18 +3169,22 @@ namespace game
...
-		    health_colour = 0xe03a1d; /* red */
-	        else if (d->health < 70)
-		    health_colour = 0xe09f1d; /* yellow */	
-		sprintf(health, "\fs\f[%d]%d%%\fS", health_colour, d->health);
-	        strcat(name_health, health);
+	        const char *name = colourname(d);
+	        int health_length;
+	        int h = d->health;
+	        for (health_length=0; h!=0; health_length++, h=h/10){}
+	        health_length += 17; /* Fixed amount of extra-characters manually calculated */
```
Comment (TheAssassin) [inline src/game/game.cpp]: There is no need for this inline comment. If you want a comment there, you should explain why this constant value is there.

File: src/game/game.cpp [hunk 6]
```diff
@@ -3169,18 +3169,22 @@ namespace game
...
+	        char name_health[strlen(name)+health_length];
```
Comment (TheAssassin) [inline src/game/game.cpp]: Variable length char arrays don't seem to be the way to go here. You should use the `MAXNAMELEN` constant instead, or just use the `string` type instead.

**Comment by Dziq84** [comment]
what ? above head - health display ? its an rpg game ? I fear this can lead to spoil game too much and even less noob friendly, like everyone will choose targets based on hp left (lets kill that one with less hp since its easier). damn, only I thought like that ?

**Comment by TheAssassin** [comment]
No @Dziq84, we discussed (and are discussing) that on IRC. I already mentioned it might lead to even more rifle camping.

File: src/game/game.cpp [hunk 7]
```diff
@@ -3170,7 +3171,7 @@ namespace game
         if(aboveheadnames && d != player1)
         {
 	    const char *name = colourname(d);
-	    int health_length = log10(INT_MAX) + 18;
+	    int health_length = MAXINTDIGITS + 18;
```
Comment (TheAssassin) [inline src/game/game.cpp]: Since this is a constant, you shouldn't calculate the sum on every run.

```
static const int health_length = MAXINTDIGITS + 18;
static const int name_health_length = MAXNAMELEN + health_length;

File: src/game/game.cpp [hunk 8]
```diff
@@ -3168,8 +3170,24 @@ namespace game
         float blend = aboveheadblend*trans;
         if(aboveheadnames && d != player1)
         {
+	    const char *name = colourname(d);
+	    static int health_length = MAXINTDIGITS + 18;
+            static const int name_health_length = MAXNAMELEN + health_length;
+	    char name_health[name_health_length];
```
Comment (shacknetisp) [inline src/game/game.cpp]: Replace this with the string type: `string name_health`

File: src/game/game.cpp [hunk 9]
```diff
@@ -3168,8 +3170,24 @@ namespace game
...
+	    strcpy(name_health, name);
```
Comment (shacknetisp) [inline src/game/game.cpp]: Here you'd use the copystring function: `copystring(name_health, name)`

File: src/game/game.cpp [hunk 10]
```diff
@@ -2,6 +2,7 @@
 #include "game.h"
 namespace game
 {
+    static const int MAXINTDIGITS = log10(INT_MAX);
```
Comment (shacknetisp) [inline src/game/game.cpp]: This line is not necessary when using the string type, which does not require a length specified.

File: src/game/game.cpp [hunk 11]
```diff
@@ -3168,8 +3170,24 @@ namespace game
...
+	    if (aboveheadhealth && (d->health > 0))
+	    {
+	        char health[health_length];
+		int health_colour = 0x45db1c; /* green */
+		if (d->health < 30)
+		    health_colour = 0xe03a1d; /* red */
+                else if (d->health < 70)
+		    health_colour = 0xe09f1d; /* yellow */	
+		sprintf(health, "\fs\f[%d]%d%%\fS", health_colour, d->health);
+	        strcat(name_health, health);
```
Comment (shacknetisp) [inline src/game/game.cpp]: Replace these two lines with the `concformatstring` function: `concformatstring(name_health, "\fs\f[%d]%d%%\fS", health_colour, d->health)`

File: src/game/game.cpp [hunk 12]
```diff
@@ -3168,8 +3170,24 @@ namespace game
         float blend = aboveheadblend*trans;
         if(aboveheadnames && d != player1)
         {
+	    const char *name = colourname(d);
+	    static int health_length = MAXINTDIGITS + 18;
+            static const int name_health_length = MAXNAMELEN + health_length;
```
Comment (shacknetisp) [inline src/game/game.cpp]: These two lines are not necessary when using the string type, which does not require a length specified.

File: src/game/game.cpp [hunk 13]
```diff
@@ -3168,8 +3170,24 @@ namespace game
...
+	    char name_health[name_health_length];
+	    strcpy(name_health, name);
+	    if (aboveheadhealth && (d->health > 0))
+	    {
+	        char health[health_length];
```
Comment (shacknetisp) [inline src/game/game.cpp]: You do not need this variable with a direct concatenation below.

**Comment by graphitemaster** [comment]
not only that formatting should be done with deffmtstring or what ever it's called

**Comment by TheAssassin** [comment]
@graphitemaster `defformatstring(...)`.

**Comment by qreeves** [comment]
The problem with the indentation is that this PR is using TAB characters whereas RE source code style uses four SPACE characters. TAB is normally eight characters long depending on how you set your viewer up, RE does not allow the use of TAB characters in the source code for indentation.

For colouring things based on an amount, there is `hud::skewcolour()`. For getting the spawn and maximum health values *for each different actor type* (bots have a different set of variables than human players) there are `m_spawnhealth()` and `m_maxhealth()` macros.

There's also an array of string functions we have in order to get around the limitations of the normal string functions and also be able to use our various string types. There's `conformatstring(dest, format, ...)` to get around needing to `sprintf()` with an intermediate variable which you'd then have to `strcat()`.

That being said, there are some concerns from the community about this topic regarding "giving away too much information", so I'd suggest the discussion be moved to the [forum topic](http://redeclipse.net/forum/viewtopic.php?f=5&t=1162) for this issue.

All in all, while I appreciate the interest in providing a feature implementation, this submission fails to follow the established style in many ways, and it's obvious you're still figuring out what everything does. Note that I'm also considering removing the whole "abovehead" system and re-implementing it screen-space rather than world-space (possibly be integrated with the new UI), so for now I think we'll close this PR.

I realise you're relatively new here as far as interacting with the development team, so please don't let the outcome of this PR discourage you from continuing to try in the future, you'll get the hang of the code in a while (I'm not a real programmer and I get by well enough with the help of everyone else). As always, I'll try to help those genuinely interested in contributing and attempt to point them in the right direction.
